### PR TITLE
Check ChainSyncClient for thunks

### PR DESCRIPTION
**Pull request by edsko**
The chain sync client stores a tip, but is polymorphic in that tip, and
if there are thunks in that tip, we might leak. For our tests we are
using the example server from the network layer, which is however not
polymorphic in its `tip` type, but uses a non-strict pair. This PR
replaces that with a strict `ExampleTip`, and pushes that through.
The `NoUnexpectedThunks` test for our chain sync tests now passes.

@coot , are you okay with the changes to the network examples here?

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 1]
```diff
@@ -796,8 +816,16 @@ rejectInvalidBlocks tracer registry getIsInvalidBlock getCandidate =
 -- the state explicit in the types and do the check in 'continueWithState'.
 newtype Stateful m blk tip s st = Stateful (s -> m (Consensus st blk tip m))
 
-continueWithState :: s -> Stateful m blk tip s st -> m (Consensus st blk tip m)
-continueWithState s (Stateful f) = f s
+continueWithState :: NoUnexpectedThunks s
+                  => s -> Stateful m blk tip s st -> m (Consensus st blk tip m)
+#if CHECK_TVAR_INVARIANT
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Thinking out loud: can't we have a single location where we have this `CPP` condition? Currently, this is the only location in `ouroboros-consensus` where we use it, but might that change?

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Let's come back to that once we have a second location :)

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: That said, if you can think of any places where we have additional implicit state..?

**Pull request merged**